components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java

```diff
@@ -447,63 +435,66 @@ private String encodePath(String path) {
     }
 
     /**
-     * Upload service Provider.
-     *
-     * @param serviceProviderDO
-     * @return
-     * @throws IdentityException
+     * Upload service Provider using metadata file..
+     * @param serviceProviderDO Service provider information object.
+     * @return True if upload success.
+     * @throws IdentityException Error occurred while adding the information to registry.
      */
     public SAMLSSOServiceProviderDO uploadServiceProvider(SAMLSSOServiceProviderDO serviceProviderDO) throws
             IdentityException {
 
-        if (serviceProviderDO.getIssuer() != null && serviceProviderDO.getDefaultAssertionConsumerUrl() != null) {
-            String path = IdentityRegistryResources.SAML_SSO_SERVICE_PROVIDERS + encodePath(serviceProviderDO
-                    .getIssuer());
-            Resource resource;
-
-            boolean isTransactionStarted = Transaction.isStarted();
-            try {
-                if (registry.resourceExists(path)) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Service Provider already exists with the same issuer name" + serviceProviderDO
-                                .getIssuer());
-                    }
-                    throw IdentityException.error("Service provider already exists");
-                }
-
-                resource = createResource(serviceProviderDO);
-
-                try {
-                    if (!isTransactionStarted) {
-                        registry.beginTransaction();
-                    }
-
-                    registry.put(path, resource);
+        if (serviceProviderDO == null || serviceProviderDO.getIssuer() == null ||
+                serviceProviderDO.getDefaultAssertionConsumerUrl() == null) {
+            throw new IdentityException("Invalid Service Provider Metadata.");
```

Comment:
Can we throw distinct exception messages for issuer = null and ACS_URL = null cases.